### ISSUE TITLE
Different resolv.conf when using "build" vs. "run"

### ISSUE THREAD

**Issue by thomaszbz** [docker/docker#15681]
I have Ubuntu 14.04.3, Kernel 3.19.0-26, Docker 1.7.0-dev, build 5e06332 from ppa http://ppa.launchpad.net/docker-maint/testing/ubuntu

My /etc/default/docker contains

    DOCKER_OPTS="--dns 8.8.8.8 --dns 192.168.178.1"

When I do a

    docker rm testdns
    docker run -it --name testdns ubuntu:14.04 cat /etc/resolv.conf; ping example.com

it works perfectly:

    nameserver 8.8.8.8
    nameserver 192.168.178.1
    search fritz.box
    PING example.com (93.184.216.34) 56(84) bytes of data.
    64 bytes from 93.184.216.34: icmp_seq=1 ttl=56 time=165 ms
    64 bytes from 93.184.216.34: icmp_seq=2 ttl=56 time=160 ms

But when I use a Dockerfile

    FROM ubuntu:14.04
    RUN     cat /etc/resolv.conf;ping example.com;ping 93.184.216.34

and do a

    docker build -t test/testdns .

it uses a different resolv.conf and DNS is not working while networking does:

    Sending build context to Docker daemon 24.06 kB
    Sending build context to Docker daemon 
    Step 0 : FROM ubuntu:14.04
     ---> 8251da35e7a7
    Step 1 : RUN cat /etc/resolv.conf;ping example.com;ping 93.184.216.34
    ---> Running in 9002497c3123
    # Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
    #     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
    nameserver 127.0.1.1
    search fritz.box
    ping: unknown host example.com
    PING 93.184.216.34 (93.184.216.34) 56(84) bytes of data.
    64 bytes from 93.184.216.34: icmp_seq=1 ttl=55 time=161 ms
    64 bytes from 93.184.216.34: icmp_seq=2 ttl=55 time=161 ms
    64 bytes from 93.184.216.34: icmp_seq=3 ttl=55 time=163 ms

Basically this means that DOCKER_OPTS DNS settings were used for "run" but not for "build". Instead, there should be no difference.

**Comment by thomaszbz** [comment]
docker version

    Client version: 1.7.0-dev
    Client API version: 1.19
    Go version (client): go1.4.2
    Git commit (client): 5e06332
    OS/Arch (client): linux/amd64
    Server version: 1.7.0-dev
    Server API version: 1.19
    Go version (server): go1.4.2
    Git commit (server): 5e06332
    OS/Arch (server): linux/amd64

docker info

    Containers: 3
    Images: 32
    Storage Driver: btrfs
    Execution Driver: native-0.2
    Logging Driver: json-file
    Kernel Version: 3.19.0-26-generic
    Operating System: Ubuntu 14.04.3 LTS
    CPUs: 4
    Total Memory: 7.6 GiB
    Name: lat61
    ID: SFHR:PHZN:6ZHO:FMPD:JBUT:U5CJ:GPE5:AI7Z:YRN3:PMWY:B24N:7NFK
    WARNING: No swap limit support

uname -a

    Linux lat61 3.19.0-26-generic #28~14.04.1-Ubuntu SMP Wed Aug 12 14:09:17 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Steps to reproduce were posted already. I also have libvirt/KVM/Qemu installed (currently no machine running).

Networking:

ifconfig -a

    docker0   Link encap:Ethernet  Hardware Adresse it:ss:pr:iv:at:ee
          inet Adresse:172.17.42.1  Bcast:0.0.0.0  Maske:255.255.0.0
          inet6-Adresse: fe80::5484:7aff:fefe:9799/64 Gültigkeitsbereich:Verbindung
          UP BROADCAST MULTICAST  MTU:1500  Metrik:1
          RX-Pakete:42 Fehler:0 Verloren:0 Überläufe:0 Fenster:0
          TX-Pakete:28 Fehler:0 Verloren:0 Überläufe:0 Träger:0
          Kollisionen:0 Sendewarteschlangenlänge:0 
          RX-Bytes:3087 (3.0 KB)  TX-Bytes:3030 (3.0 KB)

    eth0      Link encap:Ethernet  Hardware Adresse it:ss:pr:iv:at:ee  
          inet Adresse:192.168.178.26  Bcast:192.168.178.255  Maske:255.255.255.0
          inet6-Adresse: fe80::f24d:a2ff:fe90:3dc2/64 Gültigkeitsbereich:Verbindung
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metrik:1
          RX-Pakete:4851 Fehler:0 Verloren:0 Überläufe:0 Fenster:0
          TX-Pakete:5062 Fehler:0 Verloren:0 Überläufe:0 Träger:0
          Kollisionen:0 Sendewarteschlangenlänge:1000 
          RX-Bytes:2903091 (2.9 MB)  TX-Bytes:938700 (938.7 KB)
          Interrupt:17 

    lo        Link encap:Lokale Schleife  
          inet Adresse:127.0.0.1  Maske:255.0.0.0
          inet6-Adresse: ::1/128 Gültigkeitsbereich:Maschine
          UP LOOPBACK RUNNING  MTU:65536  Metrik:1
          RX-Pakete:10012 Fehler:0 Verloren:0 Überläufe:0 Fenster:0
          TX-Pakete:10012 Fehler:0 Verloren:0 Überläufe:0 Träger:0
          Kollisionen:0 Sendewarteschlangenlänge:0 
          RX-Bytes:2538278 (2.5 MB)  TX-Bytes:2538278 (2.5 MB)

    virbr0    Link encap:Ethernet  Hardware Adresse it:ss:pr:iv:at:ee  
          inet Adresse:192.168.122.1  Bcast:192.168.122.255  Maske:255.255.255.0
          UP BROADCAST MULTICAST  MTU:1500  Metrik:1
          RX-Pakete:0 Fehler:0 Verloren:0 Überläufe:0 Fenster:0
          TX-Pakete:0 Fehler:0 Verloren:0 Überläufe:0 Träger:0
          Kollisionen:0 Sendewarteschlangenlänge:0 
          RX-Bytes:0 (0.0 B)  TX-Bytes:0 (0.0 B)

    wlan0     Link encap:Ethernet  Hardware Adresse it:ss:pr:iv:at:ee  
          inet Adresse:192.168.178.22  Bcast:192.168.178.255  Maske:255.255.255.0
          inet6-Adresse: fe80::224:d7ff:fe40:dd88/64 Gültigkeitsbereich:Verbindung
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metrik:1
          RX-Pakete:350 Fehler:0 Verloren:0 Überläufe:0 Fenster:0
          TX-Pakete:71 Fehler:0 Verloren:0 Überläufe:0 Träger:0
          Kollisionen:0 Sendewarteschlangenlänge:1000 
          RX-Bytes:57141 (57.1 KB)  TX-Bytes:11608 (11.6 KB)

**Comment by thomaszbz** [comment]
I updated the same machine to Docker version 1.8.1, build d12ea79 from 

     https://apt.dockerproject.org/repo ubuntu-trusty main

And the bug is solved in this version. Test:

My /etc/default/docker contains (as before)

    DOCKER_OPTS="--dns 8.8.8.8 --dns 192.168.178.1"

My Vagrantfile is (as before)

    FROM ubuntu:14.04
    RUN     cat /etc/resolv.conf;ping example.com;ping 93.184.216.34

When I run (as before)

docker build -t test/testdns .

It returns

    Sending build context to Docker daemon 24.06 kB
    Step 0 : FROM ubuntu:14.04
     ---> 8251da35e7a7
    Step 1 : RUN cat /etc/resolv.conf;ping example.com;ping 93.184.216.34
     ---> Running in 51fd3571c4af
    nameserver 8.8.8.8
    nameserver 192.168.178.1
    search fritz.box
    PING example.com (93.184.216.34) 56(84) bytes of data.
    64 bytes from 93.184.216.34: icmp_seq=1 ttl=55 time=162 ms
    64 bytes from 93.184.216.34: icmp_seq=2 ttl=55 time=164 ms

Wich means that DOCKER_OPTS was respected as resolv.conf is corresponding. That said, DNS works and the bug has gone in 1.8.1. Should be published soon via launchpad ppa!

Issue can be closed.

**Comment by cpuguy83** [comment]
Thanks!

**Comment by thomaszbz** [comment]
For the record: With docker 1.8.1/ubuntu 14.04.3 it even works out of the box without setting DOCKER_OPTS. Just defaulting to google DNS:

    docker build -t test/testdns .

    Sending build context to Docker daemon 26.62 kB
    search fritz.box
    nameserver 8.8.8.8
    nameserver 8.8.4.4PING example.com (93.184.216.34) 56(84) bytes of data.
    64 bytes from 93.184.216.34: icmp_seq=1 ttl=55 time=160 ms
    64 bytes from 93.184.216.34: icmp_seq=2 ttl=55 time=161 ms

**Comment by trompx** [comment]
It is not working for me... Since I upgraded to docker 1.8.1, I have problem building my images :
```
Step 1 : RUN apt-get update
 ---> Running in 1c689204e4f0
Err http://archive.ubuntu.com trusty InRelease

Err http://archive.ubuntu.com trusty-updates InRelease

Err http://archive.ubuntu.com trusty-security InRelease

Err http://archive.ubuntu.com trusty Release.gpg
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com trusty-updates Release.gpg
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com trusty-security Release.gpg
  Could not resolve 'archive.ubuntu.com'
Reading package lists...
W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty/InRelease

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-updates/InRelease

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-security/InRelease

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty/Release.gpg  Could not resolve 'archive.ubuntu.com'

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-updates/Release.gpg  Could not resolve 'archive.ubuntu.com'

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-security/Release.gpg  Could not resolve 'archive.ubuntu.com'
```

When testing the same command as you
```
Step 0 : FROM ubuntu:14.04
 ---> d2a0ecffe6fa
Step 1 : RUN cat /etc/resolv.conf;ping example.com;ping 93.184.216.34
 ---> Running in 9871653779ac
nameserver 8.8.8.8
nameserver 192.168.33.105
ping: unknown host example.com
```
but I can ping just fine on the docker host.

My /etc/default/docker contains
```
DOCKER_OPTS="--dns 8.8.8.8 --dns 192.168.33.105"
```

Docker version
```
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:35:49 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:35:49 UTC 2015
 OS/Arch:      linux/amd64
```

Docker infos
```
Containers: 14
Images: 1072
Storage Driver: devicemapper
 Pool Name: docker-8:1-262623-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 8.255 GB
 Data Space Total: 107.4 GB
 Data Space Available: 30.44 GB
 Metadata Space Used: 33.51 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.114 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.77 (2012-10-15)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-52-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 2
Total Memory: 3.861 GiB
Name: test.dev
ID: F5T4:5TQR:TFFZ:WGN4:WDZT:ZX6A:ZHGY:DI4O:5GZ2:CJIV:UFLZ:TGKK
WARNING: No swap limit support
```

uname -a
```
Linux test.dev 3.13.0-52-generic #86-Ubuntu SMP Mon May 4 04:32:59 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by trompx** [comment]
I did not have those problems in 1.7.1, is it possible to set a specific version of docker-engine when we install ?

**Comment by thomaszbz** [comment]
@trompx What happens if you perform a

    docker run -it --name testdns ubuntu:14.04 cat /etc/resolv.conf; ping example.com

**Comment by trompx** [comment]
@thomaszbz The ping works
```
$ docker run -it --name testdns ubuntu:14.04 cat /etc/resolv.conf; ping example.com
Post http:///var/run/docker.sock/v1.20/containers/create?name=testdns: dial unix /var/run/docker.sock: permission denied.
* Are you trying to connect to a TLS-enabled daemon without TLS?
* Is your docker daemon up and running?
PING example.com (93.184.216.34) 56(84) bytes of data.
64 bytes from 93.184.216.34: icmp_seq=1 ttl=51 time=94.1 ms
64 bytes from 93.184.216.34: icmp_seq=2 ttl=51 time=117 ms
64 bytes from 93.184.216.34: icmp_seq=3 ttl=51 time=89.2 ms
64 bytes from 93.184.216.34: icmp_seq=4 ttl=51 time=87.7 ms
```

I tried a full system reinstall but apt-get update still not working (tried with both ubuntu:14.04 and phusion baseimage)
```
Step 0 : FROM phusion/baseimage:latest OR FROM ubuntu:14.04
 ---> e9f50c1887ea
Step 1 : RUN apt-get update
 ---> Running in 4da1b4cea753
Err http://archive.ubuntu.com trusty InRelease

Err http://archive.ubuntu.com trusty-updates InRelease

Err http://archive.ubuntu.com trusty-security InRelease

Err http://archive.ubuntu.com trusty Release.gpg
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com trusty-updates Release.gpg
  Could not resolve 'archive.ubuntu.com'
Err http://archive.ubuntu.com trusty-security Release.gpg
  Could not resolve 'archive.ubuntu.com'
Reading package lists...
W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty/InRelease

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-updates/InRelease

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-security/InRelease

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty/Release.gpg  Could not resolve 'archive.ubuntu.com'

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-updates/Release.gpg  Could not resolve 'archive.ubuntu.com'

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-security/Release.gpg  Could not resolve 'archive.ubuntu.com'

W: Some index files failed to download. They have been ignored, or old ones used instead.
```

**Comment by trompx** [comment]
On system reinstalled, I configured /etc/default/docker with
```
DOCKER_OPTS="--dns 8.8.8.8 --dns 192.168.33.105 --storage-driver=devicemapper --iptables=false"
```
thus the docker forward chain accept policy was missing.

Removing iptables=false from DOCKER_OPTS and restarting docker solved it.
